ext/nokogiri_ext_xmlsec/nokogiri_sign.c

```diff
@@ -144,23 +157,30 @@ VALUE sign_with_certificate(VALUE self, VALUE rb_opts) {
     goto done;
   }
   
-  // load certificate and add to the key
-  if(xmlSecCryptoAppKeyCertLoadMemory(dsigCtx->signKey,
-                                      (xmlSecByte *)certificate,
-                                      certificateLength,
-                                      xmlSecKeyDataFormatPem) < 0) {
-    rb_exception_result = rb_eSigningError;
-    exception_message = "failed to load certificate";
-    goto done;
+  if(keyName) {
+    // set key name
+    if(xmlSecKeySetName(dsigCtx->signKey, (xmlSecByte *)keyName) < 0) {
+      rb_exception_result = rb_eSigningError;
+      exception_message = "failed to set key name";
+      goto done;
+    }
   }
 
-  // set key name
-  if(xmlSecKeySetName(dsigCtx->signKey, (xmlSecByte *)keyName) < 0) {
-    rb_exception_result = rb_eSigningError;
-    exception_message = "failed to set key name";
-    goto done;
+  if(certificate) {
+    // load certificate and add to the key
+    if(xmlSecCryptoAppKeyCertLoadMemory(dsigCtx->signKey,
+                                        (xmlSecByte *)certificate,
+                                        certificateLength,
+                                        xmlSecKeyDataFormatPem) < 0) {
+      rb_exception_result = rb_eSigningError;
+      exception_message = "failed to load certificate";
+      goto done;
+    }
   }
 
+  // add <dsig:Signature/> node to the doc
```

Comment:
Ooh...good idea.

Heh...I hear ya on the memory management. Though I'm not ruby spoiled, I *am* C++ spoiled. Not having RAII is painful. Also, the rampant use of rb_raise() is scary too...I did a first pass cleaning this stuff up, but I just realized that any nested function call doing a rb_raise() will actually cause a leak. Ugh.